logstash-core/src/main/java/org/logstash/config/ir/compiler/ComputeStepSyntaxElement.java

```diff
@@ -33,84 +32,98 @@
         return new ComputeStepSyntaxElement<>(methods, fields, interfce);
     }
 
-    private ComputeStepSyntaxElement(final Iterable<MethodSyntaxElement> methods,
-        final ClassFields fields, final Class<T> interfce) {
+    private ComputeStepSyntaxElement(
+        final Iterable<MethodSyntaxElement> methods,
+        final ClassFields fields,
+        final Class<T> interfce)
+    {
         this.methods = methods;
         this.fields = fields;
         type = interfce;
-        classSeq = classSeqCount.incrementAndGet();
+
+        // normalizes away the name of the class so that two classes of different name but otherwise
+        // equivalent syntax get correctly compared by {@link #equals(Object)}.
+        normalizedSource = generateCode(CLASS_NAME_PLACEHOLDER);
     }
 
     @SuppressWarnings("unchecked")
-    public T instantiate(Class<? extends Dataset> clazz) {
-        try {
-            return (T) clazz.<T>getConstructor(Map.class).newInstance(ctorArguments());
+    public T instantiate() {
+         try {
+            synchronized (COMPILER) {
+                Class<? extends Dataset> clazz = CLASS_CACHE.get(this);
+                if (clazz == null) {
+                    clazz = compile();
+                    CLASS_CACHE.put(this, clazz);
+                }
+                return (T) clazz.getConstructor(Map.class).newInstance(ctorArguments());
+            }
         } catch (final NoSuchMethodException | InvocationTargetException | InstantiationException | IllegalAccessException ex) {
             throw new IllegalStateException(ex);
         }
     }
 
     @SuppressWarnings("unchecked")
-    public Class<? extends Dataset> compile() {
-        // We need to globally synchronize to avoid concurrency issues with the internal class loader
-        // Per https://github.com/elastic/logstash/pull/11482 we should review this lock. 
-        synchronized (COMPILER) {
-            try {
-                final Class<? extends Dataset> clazz;
-                final String name = String.format("CompiledDataset%d", classSeq);
-                final String code = generateCode(name);
-                if (SOURCE_DIR != null) {
-                    final Path sourceFile = SOURCE_DIR.resolve(String.format("%s.java", name));
-                    Files.write(sourceFile, code.getBytes(StandardCharsets.UTF_8));
-                    COMPILER.cookFile(sourceFile.toFile());
-                } else {
-                    COMPILER.cook(code);
-                }
-                COMPILER.setParentClassLoader(COMPILER.getClassLoader());
-                clazz = (Class<? extends Dataset>)COMPILER.getClassLoader().loadClass(
-                        String.format("org.logstash.generated.%s", name)
-                );
-
-                return clazz;
-            } catch (final CompileException | ClassNotFoundException | IOException ex) {
-                throw new IllegalStateException(ex);
+    private  Class<? extends Dataset> compile() {
```

Comment:
Let's leave a note about the need for synchronization on `COMPILER` here, maybe something like:

```suggestion
    /**
     * This method is NOT thread-safe, and must have exclusive access to `COMPILER`
     * so that the resulting `ClassLoader` after each `SimpleCompiler#cook()` operation
     * can be teed up as the parent for the next cook operation.
     */
    private  Class<? extends Dataset> compile() {
```

Reply:
agree.

Reply:
done